llvm/lib/CodeGen/PreISelIntrinsicLowering.cpp

```diff
@@ -323,7 +378,51 @@ bool PreISelIntrinsicLowering::expandMemIntrinsicUses(Function &F) const {
     }
     case Intrinsic::experimental_memset_pattern: {
       auto *Memset = cast<MemSetPatternInst>(Inst);
-      expandMemSetPatternAsLoop(Memset);
+      const TargetLibraryInfo &TLI = LookupTLI(*Memset->getFunction());
+      if (Constant *PatternValue = getMemSetPattern16Value(Memset, TLI)) {
```

Comment:
You should use early return here for the case where we have to expand via a loop.  The duplicated code in the tail is pretty minor.